In [2]:
import pandas as pd

In [3]:
data = pd.read_csv("crimeData.csv")

#### Head for the database

In [4]:
data.head()

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP DISTRICT REPORTING_AREA  \
0      I182070777           619            Larceny      C11            388   
1      I182070596           619            Larceny       C6            216   
2      I182070496           619            Larceny       B2            318   
3      I182070346           619            Larceny       B2            587   
4      I182070233           619            Larceny       B3            416   

   OCCURRED_ON_DATE        STREET        Lat       Long  \
0  23-08-2018 13:34      ADAMS ST  42.272306 -71.067214   
1  02-09-2018 17:29          F ST  42.333829 -71.051287   
2  02-09-2018 00:00     SEAVER ST  42.308789 -71.090556   
3  31-08-2018 19:00     BURNEY ST  42.331887 -71.099544   
4  31-08-2018 21:38  FAIRMOUNT ST  42.281049 -71.072064   

                        NAME  
0  LARCENY OTHER $200 & OVER  
1  LARCENY OTHER $200 & OVER  
2  LARCENY OTHER $200 & OVER  
3  LARCENY OTHER $200 & OVER  
4  LARCENY OTHER $200 & OVER

#### Tail of the Data 

In [5]:
data.tail()

INCIDENT_NUMBER  OFFENSE_CODE            OFFENSE_CODE_GROUP DISTRICT  \
108162   I140311903-00          1866                Drug Violation       D4   
108163   I140036361-00           637                       Larceny       E5   
108164   I140036361-00           639                       Larceny       E5   
108165   I130452692-00           714                    Auto Theft       B3   
108166   I130204032-00           527  Burglary - No Property Taken      E18   

       REPORTING_AREA  OCCURRED_ON_DATE               STREET        Lat  \
108162            272  07-03-2016 18:45              BALL ST  42.333489   
108163            680  10-07-2015 10:15           CORINTH ST  42.286369   
108164            680  10-07-2015 10:15           CORINTH ST  42.286369   
108165            426  05-11-2016 01:10         HARRISON AVE  42.335560   
108166            495  12-01-2016 14:45  AMERICAN LEGION HWY  42.277272   

             Long                                         NAME  
108162 -71.080642  DRUGS - POSS CLASS E INTENT TO MF DIST DISP  
108163 -71.130055              LARCENY IN A BUILDING UNDER $50  
108164 -71.130055                      LARCENY OTHER UNDER $50  
108165 -71.074364                             AUTO THEFT OTHER  
108166 -71.115730            B&E RESIDENCE DAY - NO PROP TAKEN

#### Info of the Data

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108167 entries, 0 to 108166
Data columns (total 10 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   INCIDENT_NUMBER     108167 non-null  object 
 1   OFFENSE_CODE        108167 non-null  int64  
 2   OFFENSE_CODE_GROUP  108167 non-null  object 
 3   DISTRICT            107573 non-null  object 
 4   REPORTING_AREA      108167 non-null  object 
 5   OCCURRED_ON_DATE    108167 non-null  object 
 6   STREET              104363 non-null  object 
 7   Lat                 101374 non-null  float64
 8   Long                101374 non-null  float64
 9   NAME                108167 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 8.3+ MB


#### Total missing Values in the Data (column wise)

In [7]:
data.isnull().sum()

INCIDENT_NUMBER          0
OFFENSE_CODE             0
OFFENSE_CODE_GROUP       0
DISTRICT               594
REPORTING_AREA           0
OCCURRED_ON_DATE         0
STREET                3804
Lat                   6793
Long                  6793
NAME                     0
dtype: int64

#### Total duplicates rows in the data 
#### So we have to delete all these Duplicate Data because same Incident is Store twice or more in a Database

In [8]:
data.duplicated().sum()

np.int64(21)

In [9]:
data[data.duplicated(keep=False)==True]

INCIDENT_NUMBER  OFFENSE_CODE  OFFENSE_CODE_GROUP DISTRICT  \
6555     I142049917-00          1402           Vandalism      C11   
6556     I142049917-00          1402           Vandalism      C11   
15036    I120470733-00           724          Auto Theft       C6   
15037    I120470733-00           724          Auto Theft       C6   
47269    I142020701-00           802      Simple Assault      C11   
47270    I142020701-00           802      Simple Assault      C11   
47274    I120189428-00           802      Simple Assault      E13   
47275    I120189428-00           802      Simple Assault      E13   
58397    I152030570-00          3125     Warrant Arrests      E13   
58398    I152030570-00          3125     Warrant Arrests      E13   
58399    I152026488-00          3125     Warrant Arrests       A1   
58400    I152026488-00          3125     Warrant Arrests       A1   
58404    I152004688-00          3125     Warrant Arrests       A1   
58405    I152004688-00          3125     Warrant Arrests       A1   
58415    I142017108-00          3125     Warrant Arrests       A1   
58416    I142017108-00          3125     Warrant Arrests       A1   
58417    I142014620-00          3125     Warrant Arrests       A1   
58418    I142014620-00          3125     Warrant Arrests       A1   
58423    I140056633-00          3125     Warrant Arrests       A1   
58424    I140056633-00          3125     Warrant Arrests       A1   
58425    I140036361-00          3125     Warrant Arrests       E5   
58426    I140036361-00          3125     Warrant Arrests       E5   
58429    I130371627-00          3125     Warrant Arrests       B2   
58430    I130371627-00          3125     Warrant Arrests       B2   
58432    I130202615-00          3125     Warrant Arrests       B2   
58433    I130202615-00          3125     Warrant Arrests       B2   
58439    I110177502-00          3125     Warrant Arrests       B2   
58440    I110177502-00          3125     Warrant Arrests       B2   
58891    I142017108-00          1841      Drug Violation       A1   
58892    I142017108-00          1841      Drug Violation       A1   
65075    I130194606-00           617             Larceny       D4   
65076    I130194606-00           617             Larceny       D4   
78253    I152006153-00          1849      Drug Violation       C6   
78254    I152006153-00          1849      Drug Violation       C6   
80763    I142033219-00          1842      Drug Violation       D4   
80764    I142033219-00          1842      Drug Violation       D4   
87381    I142060752-00          2610               Other       B2   
87382    I142060752-00          2610               Other       B2   
100318   I152000618-00          1503  Firearm Violations       B2   
100319   I152000618-00          1503  Firearm Violations       B2   
101508   I152041086-00          2407  Disorderly Conduct       A1   
101509   I152041086-00          2407  Disorderly Conduct       A1   

       REPORTING_AREA  OCCURRED_ON_DATE             STREET        Lat  \
6555              395  03-09-2015 13:00     DORCHESTER AVE  42.285260   
6556              395  03-09-2015 13:00     DORCHESTER AVE  42.285260   
15036             205  23-09-2016 19:15               A ST  42.346953   
15037             205  23-09-2016 19:15               A ST  42.346953   
47269             337  14-10-2015 01:00        COLUMBIA RD  42.306451   
47270             337  14-10-2015 01:00        COLUMBIA RD  42.306451   
47274             303  09-12-2015 18:35           AMORY ST  42.318245   
47275             303  09-12-2015 18:35           AMORY ST  42.318245   
58397             303  29-10-2016 08:25      W WALNUT PARK  42.317207   
58398             303  29-10-2016 08:25      W WALNUT PARK  42.317207   
58399             105  31-01-2017 10:00           FRUIT ST  42.362403   
58400             105  31-01-2017 10:00           FRUIT ST  42.362403   
58404             120  19-07-2016 12:50        BOYLSTON ST  42.352418   
58405            

#### Now We have to find out the unique values in 'Reporting_Area' Column .
#### After seeing the unique values in the data that there is a space string present in that as a unique value and Zero value also present.
#### First we find out how much values are present with space string and Zero value empty them

In [10]:
# including nan and zero 
report_area = data["REPORTING_AREA"].unique().tolist()
len(report_area)

878

In [11]:
data[(data["REPORTING_AREA"]==' ')]

INCIDENT_NUMBER  OFFENSE_CODE          OFFENSE_CODE_GROUP DISTRICT  \
15          I182068986           619                     Larceny       C6   
16          I182068695           619                     Larceny       B3   
29          I182067167           619                     Larceny       D4   
30          I182067079           619                     Larceny      C11   
31          I182066665           619                     Larceny       B2   
...                ...           ...                         ...      ...   
108088      I162030024          3029  Prisoner Related Incidents      E13   
108092      I182023530          2642     Missing Person Reported       C6   
108096      I172036311          2642     Missing Person Reported       A1   
108101      I162067604          2642     Missing Person Reported       B3   
108110      I172013557           121                Manslaughter       C6   

       REPORTING_AREA  OCCURRED_ON_DATE               STREET        Lat  \
15                     10-08-2018 12:30               PIER 4  42.351674   
16                     27-08-2018 13:54                  NaN        NaN   
29                     22-08-2018 12:20  PUBLIC ALLEY NO 543        NaN   
30                     22-08-2018 00:30                  NaN        NaN   
31                     20-08-2018 16:57            CEYLON ST        NaN   
...               ...               ...                  ...        ...   
108088                 18-04-2016 11:54                  E13        NaN   
108092                 29-03-2018 21:30                  NaN        NaN   
108096                 08-05-2017 21:17            SUMMER ST        NaN   
108101                 20-08-2016 21:38                  NaN        NaN   
108110                 19-02-2017 03:54    MASSACHUSETTS AVE        NaN   

             Long                                  NAME  
15     -71.043357             LARCENY OTHER $200 & OVER  
16            NaN             LARCENY OTHER $200 & OVER  
29            NaN             LARCENY OTHER $200 & OVER  
30            NaN             LARCENY OTHER $200 & OVER  
31            NaN             LARCENY OTHER $200 & OVER  
...           ...                                   ...  
108088        NaN  PRISONER - SUICIDE / SUICIDE ATTEMPT  
108092        NaN                     TRUANCY / RUNAWAY  
108096        NaN                     TRUANCY / RUNAWAY  
108101        NaN                     TRUANCY / RUNAWAY  
108110        NaN   MANSLAUGHTER - VEHICLE - NEGLIGENCE  

[6838 rows x 10 columns]

In [12]:
len(data[data["REPORTING_AREA"]==' '])

6838

In [13]:
data[data["REPORTING_AREA"]=='0']

INCIDENT_NUMBER  OFFENSE_CODE        OFFENSE_CODE_GROUP DISTRICT  \
11940      I172003167          3114      Investigate Property       B2   
48289      I162043274          2900                     Other       B3   
80232      I172081388          1842            Drug Violation       B2   
90701      I162015158          3116  Harbor Related Incidents      C11   

      REPORTING_AREA  OCCURRED_ON_DATE          STREET  Lat  Long  \
11940              0  12-01-2017 16:18     HILLSIDE ST  NaN   NaN   
48289              0  02-06-2016 09:02    EDGEWATER DR  NaN   NaN   
80232              0  29-09-2017 18:01     HILLSIDE ST  NaN   NaN   
90701              0  26-02-2016 07:43  DORCHESTER BAY  NaN   NaN   

                                       NAME  
11940                  INVESTIGATE PROPERTY  
48289   VAL - VIOLATION OF AUTO LAW - OTHER  
80232  DRUGS - POSS CLASS A - HEROIN, ETC.   
90701           HARBOR INCIDENT / VIOLATION

#### Now some data which have district,reportarea,and,street are all null and lat is null or (-1)
#### we have to delete all these rows because they can't be fill

In [14]:
# These values are shown when you handle the space string in the reporting area
temp = data[pd.isna(data["DISTRICT"]) & pd.isna(data["REPORTING_AREA"]) & pd.isna(data["STREET"]) & ((pd.isna(data["Lat"])) | (data["Lat"] == -1))] 
temp

Empty DataFrame
Columns: [INCIDENT_NUMBER, OFFENSE_CODE, OFFENSE_CODE_GROUP, DISTRICT, REPORTING_AREA, OCCURRED_ON_DATE, STREET, Lat, Long, NAME]
Index: []

In [15]:
len(temp)

0

#### Now we Have to check the duplicates in The "INCIDENT_NUMBER" Column 
#### There are records for multiple crime under 1 incident number this is why we are hadling those

In [16]:
duplicatedRows=data["INCIDENT_NUMBER"].duplicated(keep=False) # keep is used to select all the occurences of duplicates 
data[duplicatedRows==True]

INCIDENT_NUMBER  OFFENSE_CODE            OFFENSE_CODE_GROUP DISTRICT  \
2           I182070496           619                       Larceny       B2   
16          I182068695           619                       Larceny       B3   
18          I182068465           619                       Larceny       E5   
25          I182067539           619                       Larceny       B3   
30          I182067079           619                       Larceny      C11   
...                ...           ...                           ...      ...   
108162   I140311903-00          1866                Drug Violation       D4   
108163   I140036361-00           637                       Larceny       E5   
108164   I140036361-00           639                       Larceny       E5   
108165   I130452692-00           714                    Auto Theft       B3   
108166   I130204032-00           527  Burglary - No Property Taken      E18   

       REPORTING_AREA  OCCURRED_ON_DATE               STREET        Lat  \
2                 318  02-09-2018 00:00            SEAVER ST  42.308789   
16                     27-08-2018 13:54                  NaN        NaN   
18                691  24-08-2018 10:00            CENTRE ST  42.287094   
25                427  23-08-2018 13:54         FESSENDEN ST  42.278560   
30                     22-08-2018 00:30                  NaN        NaN   
...               ...               ...                  ...        ...   
108162            272  07-03-2016 18:45              BALL ST  42.333489   
108163            680  10-07-2015 10:15           CORINTH ST  42.286369   
108164            680  10-07-2015 10:15           CORINTH ST  42.286369   
108165            426  05-11-2016 01:10         HARRISON AVE  42.335560   
108166            495  12-01-2016 14:45  AMERICAN LEGION HWY  42.277272   

             Long                                         NAME  
2      -71.090556                    LARCENY OTHER $200 & OVER  
16            NaN                    LARCENY OTHER $200 & OVER  
18     -71.148221                    LARCENY OTHER $200 & OVER  
25     -71.091997                    LARCENY OTHER $200 & OVER  
30            NaN                    LARCENY OTHER $200 & OVER  
...           ...                                          ...  
108162 -71.080642  DRUGS - POSS CLASS E INTENT TO MF DIST DISP  
108163 -71.130055              LARCENY IN A BUILDING UNDER $50  
108164 -71.130055                      LARCENY OTHER UNDER $50  
108165 -71.074364                             AUTO THEFT OTHER  
108166 -71.115730            B&E RESIDENCE DAY - NO PROP TAKEN  

[22218 rows x 10 columns]

In [17]:
# length of the duplicated rows 
len(data[duplicatedRows==True])

22218

#### Now we have take randomly duplicates of incident number and check these incidents how they are different 
#### From below we found that 'district , reporting_area ,occured on date, street , lat and long' are same for every same incident number but 
#### Offence_Code and offence_Code_group and Name is differentiating the Incident  

In [18]:
data[data["INCIDENT_NUMBER"].isin(['I182070496','I182068465'])]

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP DISTRICT  \
2          I182070496           619            Larceny       B2   
18         I182068465           619            Larceny       E5   
1835       I182070496          1402          Vandalism       B2   
20533      I182068465          2647              Other       E5   

      REPORTING_AREA  OCCURRED_ON_DATE     STREET        Lat       Long  \
2                318  02-09-2018 00:00  SEAVER ST  42.308789 -71.090556   
18               691  24-08-2018 10:00  CENTRE ST  42.287094 -71.148221   
1835             318  02-09-2018 00:00  SEAVER ST  42.308789 -71.090556   
20533            691  24-08-2018 10:00  CENTRE ST  42.287094 -71.148221   

                            NAME  
2      LARCENY OTHER $200 & OVER  
18     LARCENY OTHER $200 & OVER  
1835                   VANDALISM  
20533  THREATS TO DO BODILY HARM

#### Now we have to check the relationship between the 'offence_code','offence_code_group' and 'name'
#### First we have to find out the unique values in these columns 

In [79]:
offenceGroup = data['OFFENSE_CODE_GROUP'].unique().tolist()
print(len(offenceGroup))

offenceCode = data['OFFENSE_CODE'].unique().tolist()
print(len(offenceCode))

name = data['NAME'].unique().tolist()
print(len(name))

66
202
202


#### So from above we found out that we find that the 
 1. No null values are present in these so all are valid 
 2. 'offence_code' and 'name' has same unique values so there are unique offence_name for every offence_code
 3. 'offence_code_group' has less unique values so there are multiple offence_code in a every offence_group

Now we have to map the (offencecode with offencecodegroup) and (offencecode with name) 

In [21]:
offenceGroupAndCode={}
offenceCodeAndName ={}
n =len(data)

for i in range(n):
    code = data["OFFENSE_CODE"].iloc[i]
    group = data["OFFENSE_CODE_GROUP"].iloc[i]
    name= data["NAME"].iloc[i]

    if group not in offenceGroupAndCode:
        offenceGroupAndCode[group] = [code]
    else:
        if code not in offenceGroupAndCode[group]:
            offenceGroupAndCode[group].append(code)

    if code not in offenceCodeAndName:
        offenceCodeAndName[code] = name
    

In [22]:
print(offenceGroupAndCode)

{'Larceny': [np.int64(619), np.int64(613), np.int64(617), np.int64(616), np.int64(612), np.int64(611), np.int64(618), np.int64(633), np.int64(627), np.int64(629), np.int64(637), np.int64(639)], 'Vandalism': [np.int64(1402), np.int64(1415)], 'Towed': [np.int64(3410)], 'Investigate Property': [np.int64(3114)], 'Motor Vehicle Accident Response': [np.int64(3820), np.int64(3801), np.int64(3803), np.int64(3831), np.int64(3802), np.int64(3807), np.int64(3830), np.int64(3810), np.int64(3821), np.int64(3811), np.int64(3805)], 'Auto Theft': [np.int64(724), np.int64(706), np.int64(727), np.int64(770), np.int64(714)], 'Verbal Disputes': [np.int64(3301)], 'Robbery': [np.int64(301), np.int64(361), np.int64(311), np.int64(381), np.int64(371), np.int64(351), np.int64(349), np.int64(339), np.int64(335)], 'Fire Related Reports': [np.int64(3108), np.int64(2612), np.int64(3160)], 'Other': [np.int64(2647), np.int64(2900), np.int64(3625), np.int64(2610), np.int64(1504), np.int64(2660), np.int64(3170), np.in

In [23]:
print(offenceCodeAndName)

{np.int64(619): 'LARCENY OTHER $200 & OVER', np.int64(1402): 'VANDALISM', np.int64(3410): 'TOWED MOTOR VEHICLE', np.int64(3114): 'INVESTIGATE PROPERTY', np.int64(3820): 'M/V ACCIDENT INVOLVING PEDESTRIAN - INJURY', np.int64(724): 'AUTO THEFT', np.int64(3301): 'VERBAL DISPUTE', np.int64(301): 'ROBBERY - STREET', np.int64(3108): 'FIRE REPORT - HOUSE, BUILDING, ETC. ', np.int64(2647): 'THREATS TO DO BODILY HARM', np.int64(3201): 'PROPERTY - LOST', np.int64(3006): 'SICK/INJURED/MEDICAL - PERSON', np.int64(3305): 'DEMONSTRATIONS/RIOT', np.int64(614): 'LARCENY NON-ACCESSORY FROM VEH. $200 & OVER', np.int64(3801): 'M/V ACCIDENT - OTHER', np.int64(3803): 'M/V ACCIDENT - PERSONAL INJURY', np.int64(522): 'B&E RESIDENCE DAY - NO FORCE', np.int64(3831): 'M/V - LEAVING SCENE - PROPERTY DAMAGE', np.int64(802): 'ASSAULT & BATTERY', np.int64(2007): 'VIOL. OF RESTRAINING ORDER W NO ARREST', np.int64(2900): 'VAL - VIOLATION OF AUTO LAW - OTHER', np.int64(2629): 'HARASSMENT', np.int64(2662): 'BALLISTICS 

#### Now we Find the unique values in 'street' columns
#### We concluded that reportingArea has lower unique values than streets (so there are one reporting area for multiple streets)  
#### So we are map the reporting area with steeets (But there are null values in both of them so we have to handle them when we do mapping) 
#### And after that we have to fill the missing values of reportarea acoording to streets

In [24]:
street = data["STREET"].unique().tolist()
print(len(street))

3713


In [25]:
street_reportNo= {}
n=len(data)
for i in range(n):
    num = data["REPORTING_AREA"].iloc[i]
    street = data["STREET"].iloc[i]
    
    if pd.isna(num) or num == ' ':
        continue
    if pd.isna(street):
        continue
    
    if num not in street_reportNo:
        street_reportNo[num] = [street]
    else:
        if street not in street_reportNo[num]:
            street_reportNo[num].append(street)    

In [26]:
street_reportNo

{'388': ['ADAMS ST',
  'DORCHESTER AVE',
  'BROOKVALE ST',
  'PATTERSON ST',
  'RICHMOND ST',
  'GALLIVAN BLVD',
  'BUTLER ST',
  'WASHINGTON ST',
  'HILLSDALE ST',
  'SWAN CT',
  'HUTCHINSON ST',
  'WESSEX ST'],
 '216': ['F ST',
  'OLD COLONY AVE',
  'D ST',
  'W SEVENTH ST',
  'W EIGHTH ST',
  'LORING ST',
  'COTTAGE ST',
  'W NINTH ST',
  'DORCHESTER ST',
  'GRIMES ST',
  'E ST',
  'BAXTER ST',
  'ALBANY ST',
  'LARK ST',
  'EARL ST',
  'GUSTIN ST',
  'MITCHELL ST',
  'FREDERICK ST',
  'EWER ST'],
 '318': ['SEAVER ST',
  'BROOKLEDGE ST',
  'HUTCHINGS ST',
  'HOMESTEAD ST',
  'ELM HILL AVE',
  'HUMBOLDT AVE',
  'GEORGIA ST'],
 '587': ['BURNEY ST',
  'TREMONT ST',
  'DELLE AVE',
  'TERRACE ST',
  'GORE ST',
  'CARMEL ST',
  'PARKER ST',
  'OSCAR ST',
  'SEWALL ST',
  'PONTIAC ST',
  'ALLEGHANY ST'],
 '416': ['FAIRMOUNT ST',
  'CODMAN HILL AVE',
  'WILMINGTON AVE',
  'FULLER ST',
  'OGDEN ST',
  'NEVADA ST',
  'CROFTLAND AVE',
  'WASHINGTON ST',
  'MILTON AVE',
  'RIVER ST'],
 '103': [

In [27]:
#so we have to fill missing values of  reportarea acoording to street
# def findReportArea(street):
#     for key,streetList in street_reportNo.items():
#         if street in streetList:
#             return key
#     return None    

# for i in range(0,n):
#     value = data.loc[i,"REPORTING_AREA"]
#     if pd.isna(value):
#         street = data.loc[i,'STREET']
#         if pd.isna(street) :
#             continue
#         data.loc[i,'REPORTING_AREA'] = findReportArea(street)

In [28]:
# data.isnull().sum()
# the null values in reportarea decreases from (6774) to (3872)

In [29]:
# After filling the missing values of report area 
# There are 503 rows are present which have street but they cant fill there reportarea because these dont have any other data which
# have reportarea so they are not mapped and all other have null values in street that is why only (2902) values reduces from reportarea.
#  data[data["STREET"].notna() & data["REPORTING_AREA"].isna()]

In [30]:
# Run the query after you reduce the report area and check 
# data[data["STREET"].notna() & data["REPORTING_AREA"].isna()]

#### Now we can fill the reportarea using (lat and long).
#### First we map the reportarea and its multiple (lat and long).
#### Then we have to fill the missing values of report area using lat and long. 

In [31]:
# We map the report area and its(lat and long)
# reportareatolatlong = {}
# for i in range(len(data)):
#     lat = data.loc[i, "Lat"]
#     long = data.loc[i, "Long"]
#     area = data.loc[i, "REPORTING_AREA"]

#     if pd.notna(lat) and pd.notna(area) and lat != -1:
#          if area not in reportareatolatlong:
#              reportareatolatlong[area] = [(lat,long)]
#          else:
#              latlonglist = reportareatolatlong[area]
#              check=True
#              for latlong in latlonglist:
#                  if latlong[0] == lat and latlong[1]==long:
#                      check=False
#                      break
#              if(check):
#                  reportareatolatlong[area].append((lat,long)) 
# reportareatolatlong        

In [32]:
# find the missing value of reportarea using the (lat and long)
# def findreportarea(lat,long):
#     for key,latLongList in reportareatolatlong.items():
#         for latlong in latLongList:
#             if lat==latlong[0] and long==latlong[1]:
#                 return key
#     return None    

# for i in range(0,n):
#     value = data.loc[i,"REPORTING_AREA"]
#     if pd.isna(value):
#         lat = data.loc[i,'Lat']
#         long = data.loc[i,'Long']
#         if pd.isna(lat) or lat==-1 :
#             continue
#         data.loc[i,'REPORTING_AREA'] = findreportarea(lat,long)

In [33]:
# (4) reportarea are filled  
# data.isnull().sum()

#### Now finding out the unique values present in the 'district' column
#### We conculded there is a relationship between the district and the reporting area for every district there are multiple reportingarea
#### So we have to map reporting area and district and there are null values also in district and reporting areas so we have to handle them 
#### And after that we have to fill the missing values of District acoording to reportarea

In [34]:
# including nan also 
districtlist = data["DISTRICT"].unique().tolist()
len(districtlist)

13

In [35]:
districtAndReportArea = {}
for i in range(n):
    district = data["DISTRICT"].iloc[i]
    reportArea = data["REPORTING_AREA"].iloc[i]

    if(pd.isna(reportArea) or reportArea == ' '):
        continue
    if(pd.isna(district)):
        continue   
    if district not in districtAndReportArea:
        districtAndReportArea[district] = [reportArea]
    else:
        if reportArea not in districtAndReportArea[district]:
            districtAndReportArea[district].append(reportArea)     
    

In [36]:
districtAndReportArea

{'C11': ['388',
  '242',
  '346',
  '366',
  '392',
  '249',
  '345',
  '450',
  '415',
  '248',
  '356',
  '355',
  '401',
  '385',
  '342',
  '335',
  '348',
  '254',
  '384',
  '351',
  '104',
  '357',
  '361',
  '359',
  '456',
  '350',
  '244',
  '373',
  '236',
  '332',
  '334',
  '823',
  '337',
  '402',
  '243',
  '362',
  '393',
  '257',
  '436',
  '253',
  '255',
  '347',
  '461',
  '338',
  '246',
  '256',
  '451',
  '387',
  '340',
  '819',
  '403',
  '360',
  '375',
  '354',
  '369',
  '363',
  '245',
  '349',
  '462',
  '394',
  '396',
  '240',
  '186',
  '455',
  '435',
  '364',
  '400',
  '352',
  '252',
  '365',
  '372',
  '935',
  '250',
  '382',
  '368',
  '397',
  '417',
  '381',
  '367',
  '398',
  '380',
  '343',
  '454',
  '336',
  '247',
  '341',
  '344',
  '370',
  '459',
  '358',
  '460',
  '235',
  '377',
  '399',
  '333',
  '395',
  '959',
  '405',
  '404',
  '406',
  '376',
  '251',
  '379',
  '374',
  '817',
  '339',
  '391',
  '378',
  '383',
  '949',
  '

In [37]:
# Now we fill the District the using the report area
# def findDistrict(reportarea):
#     for key,reportlist in districtAndReportArea.items():
#         if reportarea in reportlist:
#             return key
#     return None    

# for i in range(0,n):
#     value = data.loc[i,"DISTRICT"]
#     if pd.isna(value):
#         reportarea = data.loc[i,'REPORTING_AREA']
#         if pd.isna(reportarea) :
#             continue
#         data.loc[i,'DISTRICT'] = findDistrict(reportarea)

In [38]:
# district null values decreases from (530) values now (495)
# data.isnull().sum()

In [39]:
# Here all the missing values of districts are filled using reportarea but there are 495 district which did not have reportarea  
# data[data["REPORTING_AREA"].notna() & data["DISTRICT"].isna]

#### From above We conculded there is a relationship between the district and the street also for every district there are multiple streets
#### So we have to map district and street and there are null values also in district and street  so we have to handle them 
#### And after that we have to fill the missing values of District acoording to street

In [40]:
# Now we map the relatioship between the District and Street
districtStreets ={}
for i in range(n):
    district ,street= data.loc[i,["DISTRICT","STREET"]]

    if(pd.isna(district)):
        continue
    if(pd.isna(street)):
        continue   
        
    if district not in districtStreets:
        districtStreets[district] = [street]
    else:
        if street not in districtStreets[district]:
            districtStreets[district].append(street)     
districtStreets

{'C11': ['ADAMS ST',
  'BAKERSFIELD ST',
  'EDWIN ST',
  'ASHMONT ST',
  'SAXTON ST',
  'HOMES AVE',
  'WASHINGTON ST',
  'DRAPER ST',
  'HALLAM ST',
  'FAULKNER ST',
  'PARKMAN ST',
  'TALBOT AVE',
  'GALLIVAN BLVD',
  'BOWDOIN ST',
  'DORCHESTER AVE',
  'FREEPORT WAY',
  'HALLET ST',
  'DAKOTA ST',
  'WESTVILLE ST',
  'HOOPER ST',
  'ERIE ST',
  'ORCHARDFIELD ST',
  'MELVILLE AVE',
  'RIDGEWOOD ST',
  'HOWES ST',
  'WILLIAM T MORRISSEY BLVD',
  'EDISON GRN',
  'SAVIN HILL AVE',
  'COLUMBIA RD',
  'CLARKSON ST',
  'WINDERMERE RD',
  'OLNEY ST',
  'CLERMONT ST',
  'SAINT WILLIAM ST',
  'MAYFIELD ST',
  'MOULTRIE ST',
  'WRENTHAM ST',
  'TRENT ST',
  'EVERETT AVE',
  'PARK ST',
  'QUINCY ST',
  'TRULL ST',
  'FREEPORT ST',
  'JEROME ST',
  'JULIETTE ST',
  'SYDNEY ST',
  'LYON ST',
  'MALLON RD',
  'HANCOCK ST',
  'ASHLAND ST',
  'MOUNT IDA RD',
  'VICTORY RD',
  'FULLER ST',
  'AUCKLAND ST',
  'CHICKATAWBUT ST',
  'HOUGHTON ST',
  'EAST ST',
  'STOUGHTON ST',
  'KENWOOD ST',
  'ALLSTON

In [41]:
# now we fill the missing values of District using streets
# def findDistrict(street):
#     for key,streetList in districtStreets.items():
#         if street in streetList:
#             return key
#     return None    

# for i in range(n):
#     value = data.loc[i,"DISTRICT"]
#     if pd.isna(value):
#         street = data.loc[i,'STREET']
#         if pd.isna(street) :
#             continue
#         data.loc[i,'DISTRICT'] = findDistrict(street)

In [42]:
# district values reduce to (499) to (482)
# data.isnull().sum()

In [43]:
# There are 175 rows present where they have street but they cant fill their districts this shows that these street are not have 
# another data which have district 
# and all other are null values in street that is why we dont see any values to fill in district (only 17 values are filled) 
# data[data["STREET"].notna() & data["DISTRICT"].isna()]

#### There is a relation between district and its (lat and long)
#### First we map the DIstrict and its lat and long
#### Then we fill the missing values of district using its lat and long

In [44]:
# mapping of district using lat and long 
# districttolatlong = {}
# for i in range(len(data)):
#     lat = data.loc[i, "Lat"]
#     long = data.loc[i, "Long"]
#     district = data.loc[i, "DISTRICT"]

#     if pd.notna(lat) and pd.notna(district) and lat != -1:
#          if district not in districttolatlong:
#              districttolatlong[district] = [(lat,long)]
#          else:
#              latlonglist = districttolatlong[district]
#              check=True
#              for latlong in latlonglist:
#                  if latlong[0] == lat and latlong[1]==long:
#                      check=False
#                      break
#              if(check):
#                  districttolatlong[district].append((lat,long)) 
# districttolatlong     

In [45]:
# def finddistrict(lat,long):
#     for key,latLongList in districttolatlong.items():
#         for latlong in latLongList:
#             if lat==latlong[0] and long==latlong[1]:
#                 return key
#     return None    

# for i in range(0,n):
#     value = data.loc[i,"DISTRICT"]
#     if pd.isna(value):
#         lat = data.loc[i,'Lat']
#         long = data.loc[i,'Long']
#         if pd.isna(lat) or lat==-1 :
#             continue
#         data.loc[i,'DISTRICT'] = finddistrict(lat,long)

In [46]:
# (7) null values of district are filled from(478) to (471)
# data.isnull().sum()

#### We found that there is no any case that latitude missing only or longitude missing only 
#### There is only one case that both are missing 
#### Because total no. of (missing values in lat and long)=6793
#### And there is -1 values is present in lat and long total(220) values present shown below

In [47]:
latlongData = data[['Lat','Long']]
missing = latlongData[pd.isna(latlongData["Lat"]) & pd.isna(latlongData["Long"])]
missing

Lat  Long
16      NaN   NaN
29      NaN   NaN
30      NaN   NaN
31      NaN   NaN
49      NaN   NaN
...     ...   ...
108088  NaN   NaN
108092  NaN   NaN
108096  NaN   NaN
108101  NaN   NaN
108110  NaN   NaN

[6793 rows x 2 columns]

In [48]:
len(missing) 

6793

In [49]:
# There are  -1 values present in the lat and long present so we treated as null value 
# There are total 220 (-1) values 
latlongData[latlongData['Lat']==-1]

Lat  Long
84     -1.0  -1.0
370    -1.0  -1.0
1146   -1.0  -1.0
1591   -1.0  -1.0
1886   -1.0  -1.0
...     ...   ...
106452 -1.0  -1.0
106797 -1.0  -1.0
107268 -1.0  -1.0
107485 -1.0  -1.0
107979 -1.0  -1.0

[220 rows x 2 columns]

#### We conculded there is a relationship between the street and the (lat and long) for every street there are multiple (lat and long)
#### So we have to map street and its(lat and long) and there are null values also in street and (lat and long) so we have to handle them
#### And After that we have to fill the missing values of street acoording to (lat and long)

In [50]:
# Now we map the relationship between the street and its [lat and long]
streetLatLong = {}
for i in range(0,n):
    street = data["STREET"].iloc[i]
    lat = data["Lat"].iloc[i]
    long =  data["Long"].iloc[i]
    
    if(pd.isna(street)):
        continue
    if(pd.isna(lat) or pd.isna(long) or lat==-1 or long==-1):
        continue   
        
        
    if street not in streetLatLong:
        streetLatLong[street] =[(lat,long)]
    else:
        latlonglist = streetLatLong[street]
        check=True
        for latlong in latlonglist:
            if latlong[0] == lat and latlong[1]==long:
                check=False
                break
        if(check):
            streetLatLong[street].append((lat,long)) 
streetLatLong

{'ADAMS ST': [(np.float64(42.27230624), np.float64(-71.06721386)),
  (np.float64(42.30618185), np.float64(-71.06187015)),
  (np.float64(42.30092795), np.float64(-71.05942982)),
  (np.float64(42.30715639), np.float64(-71.06245539)),
  (np.float64(42.32802058), np.float64(-71.07716877)),
  (np.float64(42.2944367), np.float64(-71.05661774)),
  (np.float64(42.30203902), np.float64(-71.06009072)),
  (np.float64(42.29166311), np.float64(-71.05732414)),
  (np.float64(42.30636118), np.float64(-71.06195361)),
  (np.float64(42.29509311), np.float64(-71.0562477)),
  (np.float64(42.29807739), np.float64(-71.05695195)),
  (np.float64(42.30263978), np.float64(-71.06038376)),
  (np.float64(42.29760605), np.float64(-71.05676442)),
  (np.float64(42.29604546), np.float64(-71.05635478)),
  (np.float64(42.30171653), np.float64(-71.05991401)),
  (np.float64(42.29681873), np.float64(-71.05656121)),
  (np.float64(42.27113353), np.float64(-71.06819173)),
  (np.float64(42.29108353), np.float64(-71.05692998)),


In [51]:
# filling the values of (street) using (Lat and long)
# def findStreet(lat,long):
#     for key,latLongList in streetLatLong.items():
#         for latlong in latLongList:
#             if lat==latlong[0] and long==latlong[1]:
#                 return key
#     return None    

# for i in range(0,n):
#     value = data.loc[i,"STREET"]
#     if pd.isna(value):
#         lat = data.loc[i,'Lat']
#         long = data.loc[i,'Long']
#         if pd.isna(lat) or lat==-1 :
#             continue
#         data.loc[i,'STREET'] = findStreet(lat,long)

In [52]:
# No null values of Street is filled 
# data.isnull().sum()

In [53]:
# There are 671 rows present where they have lat and long but they cant fill their street this shows that these latitiude and longitude 
# are not have another data which have street.
# And all other are -1 values in and null values in lat and long that is why we dont see any values to fill null values in street
# data[data["STREET"].isna() & ((data["Lat"].notna())&(data["Lat"]!=-1))]

In [54]:
# data[data["STREET"].isna() & ((data["Lat"].notna())&(data["Lat"]!=-1))]

#### Now we have to started to fill the data on random basis.
#### because there is no data present for to fill null values and we also cant delete these rows because it affect our accuracy. 
#### There are same incident number(so there district,reportarea,street,latandlong are same) that's why we have to handle them when we fill values randomly. 

In [55]:
# filling the null values of report area using DISTRICT randomly 
# import random
# incident_to_area = {}

# for i in range(len(data)):
#     if pd.isna(data.loc[i, "REPORTING_AREA"]):
#         incident = data.loc[i, "INCIDENT_NUMBER"]

#         if incident in incident_to_area:
#             data.loc[i, "REPORTING_AREA"] = incident_to_area[incident]
#         else:
#             district = data.loc[i, "DISTRICT"]
#             if(pd.isna(district)):
#                 continue
#             possible_areas = districtAndReportArea[district]

#             if possible_areas:
#                 selected_area = random.choice(possible_areas)
#                 incident_to_area[incident] = selected_area
#                 data.loc[i, "REPORTING_AREA"] = selected_area

In [56]:
# reporting areas reduces to (3868) to (471)
# data.isnull().sum()

In [57]:
# Now we have to fill missing values of street according to reportarea randomly  
# incident_to_street= {}
# def findstreet(reportarea):
#     if reportarea in street_reportNo:
#         selected_street = random.choice(street_reportNo[reportarea])
#         return selected_street
#     return None    

# for i in range(0,n):
#     street,incident= data.loc[i,["STREET","INCIDENT_NUMBER"]]
#     if pd.isna(street):
#         if incident in incident_to_street:
#             data.loc[i, "STREET"] = incident_to_street[incident]
#         else:
#             reportarea = data.loc[i,'REPORTING_AREA']
#             if pd.isna(reportarea) :
#                 continue  
                
#             selected_street =  findstreet(reportarea)   
#             incident_to_area[incident] = selected_street    
#             data.loc[i,'STREET'] = selected_street
            

In [58]:
# street reduces to (3740) to (296) total(3436) reduces
# data.isnull().sum()

#### There is a relationship between the (lat ,long) with the unique values of (district,reportarea and,street). 
#### From below we found that there is multiple coordinates for every unique (district,reportarea and street).
#### We can fill missing values of lat and long using (district,reportarea,street) on random basis.

In [59]:
temp = data[data["DISTRICT"].notna() & data["REPORTING_AREA"].notna() & data["STREET"].notna()]
temp.reset_index(drop=True,inplace=True)

In [60]:
len(temp.groupby(["DISTRICT","REPORTING_AREA","STREET"]).size())    

8162

In [61]:
len(temp.groupby(["Lat","Long"]).size())

13743

In [62]:
#Map like this when you perform EDA and handle null in lat and long 
DRSlatlong = {}
n = len(temp)
for i in range(n):
    district,reportarea,street,lat,long = temp.loc[i,["DISTRICT","REPORTING_AREA","STREET","Lat","Long"]]
    if (district,reportarea,street) not in DRSlatlong:
        DRSlatlong[(district,reportarea,street)] = [(lat,long)]
    else:
        latlonglist = DRSlatlong[(district,reportarea,street)]
        check=True
        for latlong in latlonglist:
            if latlong[0] == lat and latlong[1]==long:
                check=False
                break
        if(check):
            DRSlatlong[(district,reportarea,street)].append((lat,long)) 
DRSlatlong

{('C11',
  '388',
  'ADAMS ST'): [(np.float64(42.27230624),
   np.float64(-71.06721386)), (np.float64(42.27282017), np.float64(-71.06525771)), (np.float64(42.27415489),
   np.float64(-71.06282889)), (np.float64(42.27616192),
   np.float64(-71.06113062)), (np.float64(42.27751943),
   np.float64(-71.06013877))],
 ('C6', '216', 'F ST'): [(np.float64(42.33382906), np.float64(-71.05128729)),
  (np.float64(42.33339546), np.float64(-71.05179708))],
 ('B2',
  '318',
  'SEAVER ST'): [(np.float64(42.30878855), np.float64(-71.09055618))],
 ('B2',
  '587',
  'BURNEY ST'): [(np.float64(42.33188687), np.float64(-71.0995439)),
  (np.float64(42.33149596), np.float64(-71.09979111))],
 ('B3',
  '416',
  'FAIRMOUNT ST'): [(np.float64(42.28104853),
   np.float64(-71.0720644)), (np.float64(42.280587), np.float64(-71.07432184))],
 ('A1',
  '103',
  'SCHOOL ST'): [(np.float64(42.3575462),
   np.float64(-71.05881953)), (np.float64(42.35774054), np.float64(-71.05956789)), (np.float64(42.35788876),
   np.float6

In [63]:
# we have to fill missing  lat and long using street or district or reporting area
# incident_to_latlong= {}
# def findlatlong(district,reportarea,street):
#     key = (district, reportarea, street)
#     if key in DRSlatlong:
#         return random.choice(DRSlatlong[key])
#     return None
   

# n = len(data)
# for i in range(n):
#     incident,district,reportarea,street,lat,long = data.loc[i, ["INCIDENT_NUMBER","DISTRICT", "REPORTING_AREA", "STREET", "Lat", "Long"]]

#     if pd.isna(lat) or lat==-1:
#         if pd.isna(district) or pd.isna(reportarea) or pd.isna(street):
#             continue
#         if incident in incident_to_latlong:
#             data.loc[i,"Lat"] = incident_to_latlong[incident][0]
#             data.loc[i,"Long"] = incident_to_latlong[incident][1] 
#         else:
#             latlong = findlatlong(district,reportarea,street)    
#             if(latlong==None):
#                 continue
#             incident_to_latlong[incident] = latlong 
#             data.loc[i, "Lat"] = latlong[0]
#             data.loc[i,"Long"] = latlong[1] 
    

In [64]:
# lat and long reduces from (6762) to (2192)
# data.isnull().sum()

#### We can fill the lat and long missing values alone by street on random basis.
#### use already made mapping between street and its (lat and long)

In [65]:
# Filling null values of lat and long using street
# for i in range(len(data)):
#     lat = data.loc[i, "Lat"]
#     street = data.loc[i, "STREET"]
    
#     if pd.isna(lat) or lat == -1:
#         if pd.isna(street):
#             continue
#         if incident in incident_to_latlong:
#             data.loc[i,"Lat"] = incident_to_latlong[incident][0]
#             data.loc[i,"Long"] = incident_to_latlong[incident][1]
#         else:
#             if street in streetLatLong:
#                 selectingLatLong = random.choice(streetLatLong[street]) 
#                 data.loc[i, "Lat"] =selectingLatLong[0] 
#                 data.loc[i, "Long"] = selectingLatLong[1]

In [66]:
## lat and long reduces from (2192) to (417)
# data.isnull().sum()

#### Now we have to remove remaining null values present in data

In [67]:
# rows_to_remove = data[
#     data["DISTRICT"].isna() |
#     data["REPORTING_AREA"].isna() |
#     data["STREET"].isna() |
#     data["Lat"].isna() 
# ]

In [68]:
# rows_to_remove

In [69]:
# len(data)
# data.drop(rows_to_remove.index,inplace=True)
# data.reset_index(drop=True,inplace=True)
# len(data)

#### Now we have to fill remaining -1 lat long values 
#### By randomly on the basis of reportarea because the street in that area have nearby lat and long present  

In [70]:
data[data["Lat"]==-1]

INCIDENT_NUMBER  OFFENSE_CODE          OFFENSE_CODE_GROUP DISTRICT  \
84          I182058509           619                     Larceny      NaN   
370         I182003973           619                     Larceny      NaN   
1146        I162073509           619                     Larceny      NaN   
1591        I152091597           619                     Larceny      A15   
1886        I182067420          1402                   Vandalism      NaN   
...                ...           ...                         ...      ...   
106452      I182003134          2619  Prisoner Related Incidents      NaN   
106797      I182061601          1845              Drug Violation      NaN   
107268      I162104752           111                    Homicide       A7   
107485      I172061061          3110    Police Service Incidents      NaN   
107979      I172067947          3122                    Aircraft      NaN   

       REPORTING_AREA  OCCURRED_ON_DATE        STREET  Lat  Long  \
84                     03-05-2018 12:44        CENTRE -1.0  -1.0   
370                    05-08-2017 18:01       FANEUIL -1.0  -1.0   
1146                   03-09-2016 15:26    HUNTINGTON -1.0  -1.0   
1591                   04-11-2015 13:55       O'MEARA -1.0  -1.0   
1886                   22-08-2018 17:30     TONAWANDA -1.0  -1.0   
...               ...               ...           ...  ...   ...   
106452                 12-01-2018 14:48      SUPERIOR -1.0  -1.0   
106797                 03-08-2018 21:08    COMMERCIAL -1.0  -1.0   
107268                 24-12-2016 22:45           NaN -1.0  -1.0   
107485                 15-07-2017 11:30  WEST ROXBURY -1.0  -1.0   
107979                 17-08-2017 20:12           NaN -1.0  -1.0   

                                       NAME  
84                LARCENY OTHER $200 & OVER  
370               LARCENY OTHER $200 & OVER  
1146              LARCENY OTHER $200 & OVER  
1591              LARCENY OTHER $200 & OVER  
1886                              VANDALISM  
...                                     ...  
106452                FUGITIVE FROM JUSTICE  
106797                 DRUGS - POSS CLASS D  
107268  MURDER, NON-NEGLIGIENT MANSLAUGHTER  
107485   SERVICE TO OTHER PD OUTSIDE OF MA.  
107979                   AIRCRAFT INCIDENTS  

[220 rows x 10 columns]

In [71]:
len(data[data["Lat"]==-1])

220

In [72]:
# filling the -1 value of lat long using report area
# n=len(data)
# def findlatlong(reportarea):
#     if reportarea in reportareatolatlong:
#         return random.choice(reportareatolatlong[reportarea])
#     return None    

# for i in range(0,n):
#     lat= data.loc[i,'Lat']
#     if lat==-1:
#         reportarea = data.loc[i,'REPORTING_AREA']
#         latlong = findlatlong(reportarea)
#         if latlong:
#             data.loc[i,'Lat'] = latlong[0]
#             data.loc[i,'Long'] = latlong[1]

In [73]:
# All values are filled 
# len(data[data["Lat"]==-1])

#### Now All the Null values are filled and -1 also filled shown below

In [74]:
# data.isnull().sum()

In [75]:
# len(data[data["Lat"]==-1])

#### Now i have to save this data 

In [76]:
# data.to_csv("crimeDataEDA.csv")